### DISCUSSION TITLE
Question on retry behavior for infinite query refetch

### DISCUSSION THREAD

**Discussion by parkyoung555** [TanStack/query#8044] [Q&A]
Hey! I'm having a hard time find an existing question on this specific scenario so hoping to get some clarity and guidance with this discussion.

I am using `useInfiniteQuery` to additively load more data into a table from a `nextToken` based list API (e.g. [example AWS SDK list API](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/sso/command/ListAccountsCommand)). Requesting next pages and refetching works great for the most part, but I'm having some trouble figuring out a way to preserve the next page param for a **retried** request triggered by a manual or automated `refetch`.

When fetching next pages with `fetchNextPage` and the `queryFn` rejects with an error response, the request is retried as expected using the same page params. However, I noticed retried requests from a manual or automated `refetch` will start over back to the first page after the first error response is encountered. I understand the need to fetch sequentially from the first page for the refetch, however, I would expect the failed request to retry using the same page params and continue onto the next page request once a retry succeeds.

The backend service I'm working with has a reasonable rate limit set (50 TPM in a 1 minute sliding window), but I'm trying to ensure a manual or automatic `refetch` will be able to recover from worst-case rate limiting scenarios where the user can continuously rate-limit themselves by refetching a table view loaded with 50+ pages.

I've put together a code sandbox with a mock API to demonstrate the page param reset behavior for retried refetch requests: https://codesandbox.io/p/sandbox/mqvdz4

Console log of requests:

```
// fetch next page

getting page 0...
got page 0
getting page 1...
got page 1
getting page 2... // simulating rate limit for page index 2
rate limited!
getting page 2...
rate limited!
getting page 2...
rate limited!
getting page 2...
got page 2 // rate limit quota restored
getting page 3...
got page 3

// refetch

getting page 0...
got page 0
getting page 1...
got page 1
getting page 2...
rate limited!
getting page 0... // page param is reset to 0, expected page 2 to be retried here
got page 0
getting page 1...
got page 1
getting page 2...
rate limited!
getting page 0... // repeatedly getting rate limited
got page 0
getting page 1...
got page 1
getting page 2...
rate limited!
getting page 0...
got page 0
getting page 1...
got page 1
getting page 2...
rate limited!
...
```

Thanks in advance for the help here! Please let me know if I need to add any more detail to this message.

**Comment by TkDodo** [comment]
thanks for reporting this. It's definitely a bug, and a serious one too. I've looked into it and I know why it's happening, but not yet sure how to fix it. Can you file an issue please?

**Reply by parkyoung555** [reply]
Appreciate the confirmation here. Created the issue https://github.com/TanStack/query/issues/8046

**Comment by parkyoung555** [comment]
Thanks for the quick turnaround here. Verified the expected behavior in v5.56.1 https://codesandbox.io/p/sandbox/tanstack-query-sandbox-forked-x77pyy.
